File: vouch/data_integrity_hybrid.py
"""
Post-quantum credentials: Ed25519 + ML-DSA-44 Data Integrity proofs.

Two shapes live here.

Proof set (the current design, per Manu Sporny's review of the W3C CG
Report, entries 9-10 of the Editor Review Queue at the top of
docs/specs/w3c-cg-report.md). The credential carries a `proof` ARRAY of two
independent proofs, `eddsa-jcs-2022` and `mldsa44-jcs-2024`. Each proof is
computed over the same unsecured document with only its own proof
configuration, and each verifies on its own, so a verifier that understands
only one of the two cryptosuites can still check that proof. Both must
verify for :func:`verify_dual` to succeed. See PAD-040 §3.3a for the
dual-proof carrier embodiment.

Composite (the v1.6.x transitional `hybrid-eddsa-mldsa44-jcs-2026`): a
single proof whose proofValue is base58btc(ed25519_sig || mldsa44_sig),
signed over the pre-alignment 32-byte digest. Verify-only: it is retained so
credentials already issued under that wire format keep verifying and so the
shared interop vector can be reproduced for regression checks. New
credentials use the proof set.

Signing input (proof set): the 64-byte W3C Data Integrity hashData,
SHA-256(canonical proof configuration) || SHA-256(canonical document). See
:func:`vouch.data_integrity.hash_data`.

Encodings:
  - `eddsa-jcs-2022` proofValue is "z" + base58btc(signature).
  - `mldsa44-jcs-2024` proofValue is "u" + base64url-nopad(signature), the
    Multibase encoding the Quantum-Resistant Cryptosuites specification uses.
    Verification also accepts the pre-alignment "z" + base58btc form and the
    pre-alignment `mldsa44-jcs-2026` identifier.

Mirrors core/vouch-core/src/hybrid.rs.

DID Document layout:
  verificationMethod[]:
    - id: did:..#key-1, type: Multikey, publicKeyMultibase: z<Ed25519>
    - id: did:..#key-2, type: Multikey, publicKeyMultibase: z<ML-DSA-44>
"""

from __future__ import annotations

import base64
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from . import data_integrity
from .multikey import _b58decode, _b58encode


# The classical cryptosuite carried alongside the post-quantum one.
CRYPTOSUITE_EDDSA = data_integrity.CRYPTOSUITE_ID
# The W3C Quantum-Resistant Cryptosuites identifier for ML-DSA-44 over JCS.
CRYPTOSUITE_MLDSA44 = "mldsa44-jcs-2024"
# Pre-alignment identifier, accepted on verification only.
CRYPTOSUITE_MLDSA44_LEGACY = "mldsa44-jcs-2026"
# The v1.6.x composite. Accepted on verification only; never emitted for new
# credentials.
CRYPTOSUITE_HYBRID_EDDSA_MLDSA44 = "hybrid-eddsa-mldsa44-jcs-2026"
PROOF_TYPE = "DataIntegrityProof"

# Fixed signature sizes for splitting the concatenated proofValue.
ED25519_SIGNATURE_SIZE = 64
MLDSA44_SIGNATURE_SIZE = 2420
HYBRID_SIGNATURE_SIZE = ED25519_SIGNATURE_SIZE + MLDSA44_SIGNATURE_SIZE


def _import_pqcrypto():
    """Import pqcrypto.sign.ml_dsa_44 lazily so the rest of the package
    keeps working when the optional dependency is missing."""
    try:
        from pqcrypto.sign import ml_dsa_44  # type: ignore[import-untyped]

        return ml_dsa_44
    except ImportError as e:
        raise ImportError(
            "Hybrid post-quantum cryptosuite requires the 'pqcrypto' package. "
            "Install it with: pip install pqcrypto"
        ) from e


def generate_mldsa44_keypair() -> Tuple[bytes, bytes]:
    """Generate a fresh ML-DSA-44 keypair. Returns (public_key, secret_key)."""
    ml_dsa_44 = _import_pqcrypto()
    return ml_dsa_44.generate_keypair()


def build_dual_proof(
    credential: Dict[str, Any],
    *,
    ed25519_private_key: Ed25519PrivateKey,
    mldsa44_secret_key: bytes,
    ed25519_verification_method: str,
    mldsa44_verification_method: Optional[str] = None,
    proof_purpose: str = "assertionMethod",
    created: Optional[datetime] = None,
) -> List[Dict[str, Any]]:
    """Build a proof set for `credential`: a list of two independent Data
    Integrity proofs, one `eddsa-jcs-2022` and one `mldsa44-jcs-2024`.

    Each proof is computed over the same unsecured document with only its own
    proof configuration, so either can be checked on its own. When
    `mldsa44_verification_method` is omitted it is derived from the Ed25519
    identifier by :func:`hybrid_verification_method_pair`.
    """
    ml_dsa_44 = _import_pqcrypto()

    base = data_integrity.unsecured_document(credential)
    created_at = created or datetime.now(timezone.utc)
    if mldsa44_verification_method is None:
        _, mldsa44_verification_method = hybrid_verification_method_pair(
            ed25519_verification_method
        )

    ed_proof = data_integrity.build_proof(
        base,
        ed25519_private_key,
        verification_method=ed25519_verification_method,
        proof_purpose=proof_purpose,
        created=created_at,
    )

    ml_proof: Dict[str, Any] = {
        "type": PROOF_TYPE,
        "cryptosuite": CRYPTOSUITE_MLDSA44,
        "created": _format_iso8601(created_at),
        "verificationMethod": mldsa44_verification_method,
        "proofPurpose": proof_purpose,
    }
    ml_signing_input = data_integrity.hash_data(base, ml_proof)
    ml_sig = ml_dsa_44.sign(mldsa44_secret_key, ml_signing_input)
    if len(ml_sig) != MLDSA44_SIGNATURE_SIZE:
        raise RuntimeError(
            f"unexpected ML-DSA-44 sig size {len(ml_sig)}, want {MLDSA44_SIGNATURE_SIZE}"
        )
    # The Quantum-Resistant Cryptosuites specification encodes proofValue as a
    # base64url-nopad Multibase value ("u"). The classical eddsa-jcs-2022 suite
    # is specified separately and keeps base58btc ("z").
    ml_proof["proofValue"] = "u" + _b64u_encode(ml_sig)

    return [ed_proof, ml_proof]


def sign_dual(
    credential: Dict[str, Any],
    *,
    ed25519_private_key: Ed25519PrivateKey,
    mldsa44_secret_key: bytes,
    ed25519_verification_method: str,
    mldsa44_verification_method: Optional[str] = None,
    proof_purpose: str = "assertionMethod",
    created: Optional[datetime] = None,
) -> Dict[str, Any]:
    """Build a proof set and return the credential with it attached under
    `proof`, replacing any existing proof."""
    proofs = build_dual_proof(
        credential,
        ed25519_private_key=ed25519_private_key,
        mldsa44_secret_key=mldsa44_secret_key,
        ed25519_verification_method=ed25519_verification_method,
        mldsa44_verification_method=mldsa44_verification_method,
        proof_purpose=proof_purpose,
        created=created,
    )
    signed = data_integrity.unsecured_document(credential)
    signed["proof"] = proofs
    return signed


def verify_dual(
    credential: Dict[str, Any],
    *,
    ed25519_public_key: Ed25519PublicKey,
    mldsa44_public_key: bytes,
) -> bool:
    """Verify a proof set: both the Ed25519 and the ML-DSA-44 proof in the
    `proof` array MUST validate. Returns True only if both are present and
    valid. Raises ValueError on a malformed proof structure."""
    ml_dsa_44 = _import_pqcrypto()

    proofs = credential.get("proof")
    if not isinstance(proofs, list):
        raise ValueError("A dual proof requires a proof array")
    base = data_integrity.unsecured_document(credential)

    # Every recognized proof in the set must verify, not just one of each kind,
    # so a set carrying a good proof next to a bad one is rejected.
    ed_ok = False
    ml_ok = False
    for proof in proofs:
        if not isinstance(proof, dict):
            continue
        cryptosuite = proof.get("cryptosuite")
        if cryptosuite == CRYPTOSUITE_EDDSA:
            candidate = dict(base)
            candidate["proof"] = proof
            if not data_integrity.verify_proof(candidate, ed25519_public_key):
                return False
            ed_ok = True
        elif cryptosuite in (CRYPTOSUITE_MLDSA44, CRYPTOSUITE_MLDSA44_LEGACY):
            proof_value = proof.get("proofValue")
            if not isinstance(proof_value, str):
                raise ValueError("ML-DSA proof missing proofValue")
            # Accept the specified base64url-nopad encoding, and the
            # pre-alignment base58btc encoding for credentials already issued.
            if proof_value.startswith("u"):
                ml_sig = _b64u_decode(proof_value[1:])
            elif proof_value.startswith("z"):
                ml_sig = _b58decode(proof_value[1:])
            else:
                raise ValueError("proofValue must be multibase base64url (u) or base58btc (z)")
            unsigned = {k: v for k, v in proof.items() if k != "proofValue"}
            verified = _mldsa44_verify(
                ml_dsa_44,
                mldsa44_public_key,
                data_integrity.hash_data(base, unsigned),
                ml_sig,
            )
            if not verified:
                # Fall back to the pre-alignment signing input.
                verified = _mldsa44_verify(
                    ml_dsa_44,
                    mldsa44_public_key,
                    data_integrity.legacy_proof_digest(base, unsigned),
                    ml_sig,
                )
            if not verified:
                return False
            ml_ok = True
    # Both members must be PRESENT: a set carrying only the classical proof does
    # not verify, so a credential cannot silently drop the post-quantum proof.
    return ed_ok and ml_ok


def build_hybrid_proof(
    credential: Dict[str, Any],
    *,
    ed25519_private_key: Ed25519PrivateKey,
    mldsa44_secret_key: bytes,
    verification_method: str,
    proof_purpose: str = "assertionMethod",
    created: Optional[datetime] = None,
) -> Dict[str, Any]:
    """Build a v1.6.x composite proof (a single proof whose proofValue is
    base58btc(ed25519_sig || mldsa44_sig)) for `credential`, using the
    pre-alignment signing input that format was issued under.

    Verify-only wire format: retained so the older shape can be reproduced for
    regression checks against credentials issued under v1.6.x. New credentials
    use :func:`build_dual_proof`, which emits a proof set.
    """
    ml_dsa_44 = _import_pqcrypto()

    proof: Dict[str, Any] = {
        "type": PROOF_TYPE,
        "cryptosuite": CRYPTOSUITE_HYBRID_EDDSA_MLDSA44,
        "created": _format_iso8601(created or datetime.now(timezone.utc)),
        "verificationMethod": verification_method,
        "proofPurpose": proof_purpose,
    }

    # The composite format signs the pre-alignment 32-byte digest.
    digest = data_integrity.legacy_proof_digest(credential, proof)

    # Classical Ed25519 signature.
    ed_sig = ed25519_private_key.sign(digest)
    if len(ed_sig) != ED25519_SIGNATURE_SIZE:
        raise RuntimeError(
            f"unexpected Ed25519 sig size {len(ed_sig)}, want {ED25519_SIGNATURE_SIZE}"
        )

    # Post-quantum ML-DSA-44 signature.
    ml_sig = ml_dsa_44.sign(mldsa44_secret_key, digest)
    if len(ml_sig) != MLDSA44_SIGNATURE_SIZE:
        raise RuntimeError(
            f"unexpected ML-DSA-44 sig size {len(ml_sig)}, want {MLDSA44_SIGNATURE_SIZE}"
        )

    combined = ed_sig + ml_sig
    proof["proofValue"] = "z" + _b58encode(combined)
    return proof


def verify_hybrid_proof(
    credential: Dict[str, Any],
    *,
    ed25519_public_key: Ed25519PublicKey,
    mldsa44_public_key: bytes,
) -> bool:
    """Verify a v1.6.x composite proof (single proof, concatenated proofValue)
    against the pre-alignment signing input that format was issued under.

    Both signatures MUST validate. Returns True on success, False on
    signature failure. Raises ValueError on malformed proof structure.
    """
    ml_dsa_44 = _import_pqcrypto()

    proof = credential.get("proof")
    if not isinstance(proof, dict):
        raise ValueError("Credential has no proof object")
    if proof.get("type") != PROOF_TYPE:
        raise ValueError(f"Unexpected proof type: {proof.get('type')}")
    if proof.get("cryptosuite") != CRYPTOSUITE_HYBRID_EDDSA_MLDSA44:
        raise ValueError(f"Unexpected cryptosuite: {proof.get('cryptosuite')}")

    proof_value = proof.get("proofValue")
    if not isinstance(proof_value, str) or not proof_value.startswith("z"):
        raise ValueError("Missing or malformed proofValue")

    combined = _b58decode(proof_value[1:])
    if len(combined) != HYBRID_SIGNATURE_SIZE:
        raise ValueError(
            f"hybrid signature length {len(combined)}, expected {HYBRID_SIGNATURE_SIZE}"
        )

    ed_sig = combined[:ED25519_SIGNATURE_SIZE]
    ml_sig = combined[ED25519_SIGNATURE_SIZE:]

    # Reconstruct the pre-alignment signing input.
    proof_without_value = {k: v for k, v in proof.items() if k != "proofValue"}
    digest = data_integrity.legacy_proof_digest(credential, proof_without_value)

    # Classical Ed25519 verification.
    try:
        ed25519_public_key.verify(ed_sig, digest)
    except InvalidSignature:
        return False

    # Post-quantum ML-DSA-44 verification.
    return _mldsa44_verify(ml_dsa_44, mldsa44_public_key, digest, ml_sig)


def hybrid_verification_method_pair(verification_method: str) -> Tuple[str, str]:
    """Derive the (Ed25519, ML-DSA-44) verificationMethod URL pair from a
    single identifier. The convention is that the proof's verificationMethod
    points at the Ed25519 key (#key-1) and the ML-DSA-44 key sits at the
    parallel slot (#key-2) on the same DID."""
    if verification_method.endswith("#key-1"):
        base = verification_method[: -len("#key-1")]
        return verification_method, base + "#key-2"
    if "#" in verification_method:
        idx = verification_method.index("#")
        return verification_method, verification_method[:idx] + "#key-2"
    return verification_method, verification_method + "#key-2"


def _b64u_encode(raw: bytes) -> str:
    """Multibase base64url-nopad body."""
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def _b64u_decode(body: str) -> bytes:
    padding = "=" * (-len(body) % 4)
    try:
        return base64.urlsafe_b64decode(body + padding)
    except Exception as exc:
        raise ValueError(f"bad base64url proofValue: {exc}") from exc


def _mldsa44_verify(ml_dsa_44: Any, public_key: bytes, message: bytes, signature: bytes) -> bool:
    # pqcrypto's verify raises on an invalid signature in some builds; treat
    # any exception as a verification failure.
    try:
        return bool(ml_dsa_44.verify(public_key, message, signature))
    except Exception:
        return False


def _format_iso8601(dt: datetime) -> str:
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

File: packages/sdk-ts/src/data-integrity-hybrid.ts
/**
 * Post-quantum credentials as a Data Integrity proof set (TypeScript).
 *
 * The current shape is a `proof` ARRAY of two independent proofs,
 * `eddsa-jcs-2022` and `mldsa44-jcs-2024`. Each proof is computed over the
 * same unsecured document with only its own proof configuration, and each
 * verifies on its own, so a verifier that understands only one of the two
 * cryptosuites can still check that proof. Both must verify for
 * {@link verifyDualProof} to succeed. See PAD-040 §3.3a for the dual-proof
 * carrier embodiment.
 *
 * Two pre-alignment shapes are accepted on verification: the
 * `mldsa44-jcs-2026` identifier with a base58btc proofValue, and the v1.6.x
 * composite `hybrid-eddsa-mldsa44-jcs-2026` whose single proofValue was
 * base58btc(ed25519_sig || mldsa44_sig). The composite builder is retained
 * only so the older wire format can be reproduced for the credentials already
 * in the field (fielded robots re-signed under the v1.6.x profile).
 *
 * Mirrors core/vouch-core/src/hybrid.rs, go-sidecar/signer and the Python
 * counterpart. The wire format is identical across implementations so that
 * a credential signed by one can be verified by another.
 */

import * as crypto from 'crypto';
import { createRequire } from 'node:module';

import { b58decode, b58encode } from './multikey';
import {
  buildProof,
  verifyProof,
  hashData,
  legacyProofDigest,
  CRYPTOSUITE_ID as EDDSA_CRYPTOSUITE_ID,
  PROOF_TYPE,
} from './data-integrity';
import type { DataIntegrityProof } from './data-integrity';

// @noble/post-quantum is an OPTIONAL peer dependency. It is loaded lazily on
// first use (not at module load) so that importing the SDK never requires it
// to be installed, and so the package imports cleanly under ESM. We pin to
// ^0.4 because that line ships a CJS build resolvable via `createRequire`.
//
// API note: the 0.4.x signature is `sign(secretKey, msg)` and
// `verify(publicKey, msg, sig)`. The 0.6.x line moved to `sign(msg, secretKey)`
// and `verify(sig, msg, publicKey)`. We follow 0.4.x here. Wire format
// (raw signature bytes) is identical across versions.
interface MlDsa44 {
  keygen(seed?: Uint8Array): { secretKey: Uint8Array; publicKey: Uint8Array };
  sign(secretKey: Uint8Array, msg: Uint8Array): Uint8Array;
  verify(publicKey: Uint8Array, msg: Uint8Array, sig: Uint8Array): boolean;
}

let _mlDsa44: MlDsa44 | undefined;

/**
 * Lazily resolve the ML-DSA-44 implementation from the optional
 * `@noble/post-quantum` peer dependency. Uses `createRequire` so it works in
 * both the CJS and ESM builds (esbuild rewrites a bare `require()` into a
 * throwing shim in ESM output). Throws a clear, actionable error if the
 * optional dependency is not installed.
 */
function getMlDsa44(): MlDsa44 {
  if (_mlDsa44) return _mlDsa44;
  let mod: { ml_dsa44?: MlDsa44; default?: { ml_dsa44?: MlDsa44 } };
  try {
    const req = createRequire(import.meta.url);
    mod = req('@noble/post-quantum/ml-dsa');
  } catch (err) {
    throw new Error(
      'The hybrid post-quantum profile requires the optional peer dependency ' +
      '"@noble/post-quantum" (^0.4). Install it with `npm install @noble/post-quantum` ' +
      'to use hybrid Ed25519 + ML-DSA-44 proofs. Original error: ' +
      (err instanceof Error ? err.message : String(err))
    );
  }
  const impl = mod.ml_dsa44 ?? mod.default?.ml_dsa44;
  if (!impl) {
    throw new Error('@noble/post-quantum/ml-dsa did not export `ml_dsa44`.');
  }
  _mlDsa44 = impl;
  return _mlDsa44;
}

/** The v1.6.x composite. Accepted on verification only; never emitted. */
export const HYBRID_CRYPTOSUITE_ID = 'hybrid-eddsa-mldsa44-jcs-2026';
/** The W3C Quantum-Resistant Cryptosuites identifier for ML-DSA-44 over JCS. */
export const MLDSA44_CRYPTOSUITE_ID = 'mldsa44-jcs-2024';
/** Pre-alignment ML-DSA identifier, accepted on verification only. */
export const MLDSA44_LEGACY_CRYPTOSUITE_ID = 'mldsa44-jcs-2026';
export { EDDSA_CRYPTOSUITE_ID };

const ED25519_SIG_SIZE = 64;
const MLDSA44_SIG_SIZE = 2420;
const HYBRID_SIG_SIZE = ED25519_SIG_SIZE + MLDSA44_SIG_SIZE;

export interface HybridKeyPair {
  secretKey: Uint8Array;
  publicKey: Uint8Array;
}

/**
 * Generate a fresh ML-DSA-44 keypair via @noble/post-quantum.
 */
export function generateMLDSA44KeyPair(seed?: Uint8Array): HybridKeyPair {
  const ml_dsa44 = getMlDsa44();
  const { secretKey, publicKey } = ml_dsa44.keygen(seed);
  return { secretKey, publicKey };
}

export interface BuildHybridProofOptions {
  ed25519PrivateKey: crypto.KeyObject;
  mldsa44SecretKey: Uint8Array;
  verificationMethod: string;
  proofPurpose?: string;
  created?: Date;
}

/**
 * Build a v1.6.x composite proof (a single proof whose proofValue is
 * base58btc(ed25519_sig || mldsa44_sig)), using the pre-alignment signing
 * input that format was issued under: the SHA-256 digest of the
 * JCS-canonicalized credential with the unsigned proof attached.
 *
 * Retained so the older wire format stays reproducible for the credentials
 * already in the field. New credentials use {@link buildDualProof}, which
 * emits a proof set.
 *
 * @deprecated the composite wire format is verify-only; use buildDualProof
 */
export function buildHybridProof(
  credential: Record<string, unknown>,
  opts: BuildHybridProofOptions
): DataIntegrityProof {
  const proof: DataIntegrityProof = {
    type: PROOF_TYPE,
    cryptosuite: HYBRID_CRYPTOSUITE_ID,
    created: formatIso8601(opts.created ?? new Date()),
    verificationMethod: opts.verificationMethod,
    proofPurpose: opts.proofPurpose ?? 'assertionMethod',
  };

  const digest = legacyProofDigest(
    stripProof(credential),
    proof as unknown as Record<string, unknown>
  );

  // Ed25519: use Node crypto (algorithm null for Ed25519 keys).
  const edSig = crypto.sign(null, digest, opts.ed25519PrivateKey);
  if (edSig.length !== ED25519_SIG_SIZE) {
    throw new Error(`unexpected Ed25519 sig size ${edSig.length}`);
  }

  // ML-DSA-44 via @noble/post-quantum (0.4.x: secretKey, msg order).
  const mlSig = getMlDsa44().sign(opts.mldsa44SecretKey, digest);
  if (mlSig.length !== MLDSA44_SIG_SIZE) {
    throw new Error(`unexpected ML-DSA-44 sig size ${mlSig.length}`);
  }

  const combined = new Uint8Array(HYBRID_SIG_SIZE);
  combined.set(edSig, 0);
  combined.set(mlSig, ED25519_SIG_SIZE);

  proof.proofValue = 'z' + b58encode(combined);
  return proof;
}

/**
 * Verify a v1.6.x composite hybrid proof (single proof, concatenated
 * proofValue) against the pre-alignment signing input it was issued under.
 * Both embedded signatures MUST validate. Returns true on success, false on
 * signature failure. Throws on malformed proof structure.
 */
export function verifyHybridProof(
  credential: Record<string, unknown>,
  ed25519PublicKey: crypto.KeyObject,
  mldsa44PublicKey: Uint8Array
): boolean {
  const proof = (credential as { proof?: DataIntegrityProof }).proof;
  if (!proof || typeof proof !== 'object') {
    throw new Error('Credential has no proof object');
  }
  if (proof.type !== PROOF_TYPE) {
    throw new Error(`Unexpected proof type: ${proof.type}`);
  }
  if (proof.cryptosuite !== HYBRID_CRYPTOSUITE_ID) {
    throw new Error(`Unexpected cryptosuite: ${proof.cryptosuite}`);
  }
  if (!proof.proofValue || !proof.proofValue.startsWith('z')) {
    throw new Error('Missing or malformed proofValue');
  }

  const combined = b58decode(proof.proofValue.slice(1));
  if (combined.length !== HYBRID_SIG_SIZE) {
    throw new Error(
      `hybrid signature length ${combined.length}, expected ${HYBRID_SIG_SIZE}`
    );
  }

  const edSig = combined.slice(0, ED25519_SIG_SIZE);
  const mlSig = combined.slice(ED25519_SIG_SIZE);

  const proofWithoutValue: Partial<DataIntegrityProof> = { ...proof };
  delete proofWithoutValue.proofValue;
  const digest = legacyProofDigest(
    stripProof(credential),
    proofWithoutValue as unknown as Record<string, unknown>
  );

  if (!crypto.verify(null, digest, ed25519PublicKey, Buffer.from(edSig))) {
    return false;
  }
  // 0.4.x order: publicKey, msg, sig
  if (!getMlDsa44().verify(mldsa44PublicKey, digest, mlSig)) {
    return false;
  }
  return true;
}

/** Strip any `proof` member, yielding the unsecured document. */
function stripProof(credential: Record<string, unknown>): Record<string, unknown> {
  const base: Record<string, unknown> = { ...credential };
  delete base.proof;
  return base;
}

/** Multibase base64url-nopad ("u") encoding, per the PQ cryptosuites. */
function mb64u(bytes: Uint8Array): string {
  return 'u' + Buffer.from(bytes).toString('base64url');
}

/** Decode a proofValue that is either base64url-nopad ("u") or base58btc ("z"). */
function decodeProofValue(proofValue: string): Uint8Array {
  if (proofValue.startsWith('u')) {
    return new Uint8Array(Buffer.from(proofValue.slice(1), 'base64url'));
  }
  if (proofValue.startsWith('z')) {
    return b58decode(proofValue.slice(1));
  }
  throw new Error('proofValue must be multibase base64url (u) or base58btc (z)');
}

export interface BuildDualProofOptions {
  ed25519PrivateKey: crypto.KeyObject;
  mldsa44SecretKey: Uint8Array;
  /** verificationMethod for the Ed25519 proof (conventionally #key-1). */
  ed25519VerificationMethod: string;
  /**
   * verificationMethod for the ML-DSA-44 proof. When omitted it is derived
   * from the Ed25519 identifier by {@link hybridVerificationMethodPair} (the
   * #key-2 slot on the same DID).
   */
  mldsa44VerificationMethod?: string;
  proofPurpose?: string;
  created?: Date;
}

/**
 * Build a dual proof: an ARRAY of two independent Data Integrity proofs,
 * `eddsa-jcs-2022` and `mldsa44-jcs-2024`, over the same unsecured document.
 * Each proof stands alone and both must verify. The ML-DSA proofValue is the
 * multibase base64url-nopad ("u") encoding specified by the Quantum-Resistant
 * Cryptosuites; the classical suite keeps base58btc ("z").
 */
export function buildDualProof(
  credential: Record<string, unknown>,
  opts: BuildDualProofOptions
): DataIntegrityProof[] {
  const base = stripProof(credential);
  const created = opts.created ?? new Date();
  const proofPurpose = opts.proofPurpose ?? 'assertionMethod';
  const mldsa44VerificationMethod =
    opts.mldsa44VerificationMethod ??
    hybridVerificationMethodPair(opts.ed25519VerificationMethod).mldsa44;

  const edProof = buildProof(base, {
    privateKey: opts.ed25519PrivateKey,
    verificationMethod: opts.ed25519VerificationMethod,
    proofPurpose,
    created,
  });

  const mlProof: DataIntegrityProof = {
    type: PROOF_TYPE,
    cryptosuite: MLDSA44_CRYPTOSUITE_ID,
    created: formatIso8601(created),
    verificationMethod: mldsa44VerificationMethod,
    proofPurpose,
  };
  const signingInput = hashData(base, mlProof as unknown as Record<string, unknown>);
  const mlSig = getMlDsa44().sign(opts.mldsa44SecretKey, signingInput);
  if (mlSig.length !== MLDSA44_SIG_SIZE) {
    throw new Error(`unexpected ML-DSA-44 sig size ${mlSig.length}`);
  }
  mlProof.proofValue = mb64u(mlSig);

  return [edProof, mlProof];
}

/**
 * Build a dual proof and attach it to the credential under `proof`, replacing
 * any existing proof.
 */
export function signDual(
  credential: Record<string, unknown>,
  opts: BuildDualProofOptions
): Record<string, unknown> {
  return { ...stripProof(credential), proof: buildDualProof(credential, opts) };
}

/**
 * True if `proof` is a proof set carrying an ML-DSA-44 proof (either the
 * specified `mldsa44-jcs-2024` identifier or the pre-alignment one).
 */
export function isMlDsaCryptosuite(cryptosuite: unknown): boolean {
  return (
    cryptosuite === MLDSA44_CRYPTOSUITE_ID ||
    cryptosuite === MLDSA44_LEGACY_CRYPTOSUITE_ID
  );
}

/**
 * Verify a dual proof: both the Ed25519 and the ML-DSA-44 proof in the array
 * MUST validate. Returns true only if both are present and every proof in the
 * set validates. Accepts the pre-alignment `mldsa44-jcs-2026` identifier, a
 * base58btc ML-DSA proofValue, and the pre-alignment signing input.
 *
 * SECURITY: no proof in the set is ever skipped. A proof carrying a
 * cryptosuite this function cannot check throws rather than being ignored, and
 * a single failing member fails the whole set.
 */
export function verifyDualProof(
  credential: Record<string, unknown>,
  ed25519PublicKey: crypto.KeyObject,
  mldsa44PublicKey: Uint8Array
): boolean {
  const proofs = (credential as { proof?: unknown }).proof;
  if (!Array.isArray(proofs)) {
    throw new Error('dual proof requires a proof array');
  }
  const base = stripProof(credential);

  let edSeen = false;
  let mlSeen = false;
  let allOk = true;

  for (const entry of proofs as DataIntegrityProof[]) {
    const cryptosuite = entry?.cryptosuite;
    if (cryptosuite === EDDSA_CRYPTOSUITE_ID) {
      edSeen = true;
      allOk = verifyProof({ ...base, proof: entry }, ed25519PublicKey) && allOk;
    } else if (isMlDsaCryptosuite(cryptosuite)) {
      mlSeen = true;
      if (!entry.proofValue) {
        throw new Error('ml-dsa proof missing proofValue');
      }
      const sig = decodeProofValue(entry.proofValue);
      const unsigned: Partial<DataIntegrityProof> = { ...entry };
      delete unsigned.proofValue;
      const unsignedProof = unsigned as unknown as Record<string, unknown>;

      const ml = getMlDsa44();
      let mlOk = ml.verify(mldsa44PublicKey, hashData(base, unsignedProof), sig);
      if (!mlOk) {
        mlOk = ml.verify(
          mldsa44PublicKey,
          legacyProofDigest(base, unsignedProof),
          sig
        );
      }
      allOk = mlOk && allOk;
    }
    // A proof for an unrecognized cryptosuite is ignored, so a proof set may
    // carry an additional future cryptosuite without breaking a verifier that
    // predates it. This is safe because acceptance still requires both known
    // members to be present and to verify; an unrecognized proof can never
    // stand in for a missing or failing known proof.
  }
  return edSeen && mlSeen && allOk;
}

/**
 * Derive the (#key-1, #key-2) verificationMethod pair from a single
 * verificationMethod identifier. The convention is that the proof's
 * verificationMethod points at the Ed25519 key (#key-1) and the
 * ML-DSA-44 key sits at the parallel slot (#key-2) on the same DID.
 */
export function hybridVerificationMethodPair(
  verificationMethod: string
): { ed25519: string; mldsa44: string } {
  if (verificationMethod.endsWith('#key-1')) {
    const base = verificationMethod.slice(0, -'#key-1'.length);
    return { ed25519: verificationMethod, mldsa44: base + '#key-2' };
  }
  const hashIdx = verificationMethod.indexOf('#');
  if (hashIdx >= 0) {
    return {
      ed25519: verificationMethod,
      mldsa44: verificationMethod.slice(0, hashIdx) + '#key-2',
    };
  }
  return {
    ed25519: verificationMethod,
    mldsa44: verificationMethod + '#key-2',
  };
}

function formatIso8601(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}
